File: binding/src/charts/libqlogvalueaxis.cpp
#include <QAbstractAxis>
#include <QLogValueAxis>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qlogvalueaxis.h>
#include "libqlogvalueaxis.h"
#include "libqlogvalueaxis.hxx"

QLogValueAxis* QLogValueAxis_new() {
	 return new VirtualQLogValueAxis();
}

QLogValueAxis* QLogValueAxis_new2(QObject* parent) {
	 return new VirtualQLogValueAxis(parent);
}

libqt_string QLogValueAxis_Tr(const char* s) {
	QString _ret = QLogValueAxis::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QLogValueAxis_Type(const QLogValueAxis* self) {
	return self->type();
}

void QLogValueAxis_SetMin(QLogValueAxis* self, double min) {
	self->setMin(min);
}

double QLogValueAxis_Min(const QLogValueAxis* self) {
	return self->min();
}

void QLogValueAxis_SetMax(QLogValueAxis* self, double max) {
	self->setMax(max);
}

double QLogValueAxis_Max(const QLogValueAxis* self) {
	return self->max();
}

void QLogValueAxis_SetRange(QLogValueAxis* self, double min, double max) {
	self->setRange(min, max);
}

void QLogValueAxis_SetLabelFormat(QLogValueAxis* self, const libqt_string format) {
	self->setLabelFormat(QString::fromUtf8(format.data, format.len));
}

libqt_string QLogValueAxis_LabelFormat(const QLogValueAxis* self) {
	QString _ret = self->labelFormat();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QLogValueAxis_SetBase(QLogValueAxis* self, double base) {
	self->setBase(base);
}

double QLogValueAxis_Base(const QLogValueAxis* self) {
	return self->base();
}

int QLogValueAxis_TickCount(const QLogValueAxis* self) {
	return self->tickCount();
}

void QLogValueAxis_SetMinorTickCount(QLogValueAxis* self, int minorTickCount) {
	self->setMinorTickCount(minorTickCount);
}

int QLogValueAxis_MinorTickCount(const QLogValueAxis* self) {
	return self->minorTickCount();
}

void QLogValueAxis_MinChanged(QLogValueAxis* self, double min) {
	self->minChanged(min);
}

void QLogValueAxis_Connect_MinChanged(QLogValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QLogValueAxis*, double) = reinterpret_cast<void (*)(QLogValueAxis*, double)>(slot);
    QLogValueAxis::connect(self, &QLogValueAxis::minChanged, [self, slotFunc](qreal min) {
	slotFunc(self, min);
    });
}

void QLogValueAxis_MaxChanged(QLogValueAxis* self, double max) {
	self->maxChanged(max);
}

void QLogValueAxis_Connect_MaxChanged(QLogValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QLogValueAxis*, double) = reinterpret_cast<void (*)(QLogValueAxis*, double)>(slot);
    QLogValueAxis::connect(self, &QLogValueAxis::maxChanged, [self, slotFunc](qreal max) {
	slotFunc(self, max);
    });
}

void QLogValueAxis_RangeChanged(QLogValueAxis* self, double min, double max) {
	self->rangeChanged(min, max);
}

void QLogValueAxis_Connect_RangeChanged(QLogValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QLogValueAxis*, double, double) = reinterpret_cast<void (*)(QLogValueAxis*, double, double)>(slot);
    QLogValueAxis::connect(self, &QLogValueAxis::rangeChanged, [self, slotFunc](qreal min, qreal max) {
	slotFunc(self, min, max);
    });
}

void QLogValueAxis_LabelFormatChanged(QLogValueAxis* self, const libqt_string format) {
	self->labelFormatChanged(QString::fromUtf8(format.data, format.len));
}

void QLogValueAxis_Connect_LabelFormatChanged(QLogValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QLogValueAxis*, const char*) = reinterpret_cast<void (*)(QLogValueAxis*, const char*)>(slot);
    QLogValueAxis::connect(self, &QLogValueAxis::labelFormatChanged, [self, slotFunc](const QString& format) {
	slotFunc(self, format);
    });
}

void QLogValueAxis_BaseChanged(QLogValueAxis* self, double base) {
	self->baseChanged(base);
}

void QLogValueAxis_Connect_BaseChanged(QLogValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QLogValueAxis*, double) = reinterpret_cast<void (*)(QLogValueAxis*, double)>(slot);
    QLogValueAxis::connect(self, &QLogValueAxis::baseChanged, [self, slotFunc](qreal base) {
	slotFunc(self, base);
    });
}

void QLogValueAxis_TickCountChanged(QLogValueAxis* self, int tickCount) {
	self->tickCountChanged(tickCount);
}

void QLogValueAxis_Connect_TickCountChanged(QLogValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QLogValueAxis*, int) = reinterpret_cast<void (*)(QLogValueAxis*, int)>(slot);
    QLogValueAxis::connect(self, &QLogValueAxis::tickCountChanged, [self, slotFunc](int tickCount) {
	slotFunc(self, tickCount);
    });
}

void QLogValueAxis_MinorTickCountChanged(QLogValueAxis* self, int minorTickCount) {
	self->minorTickCountChanged(minorTickCount);
}

void QLogValueAxis_Connect_MinorTickCountChanged(QLogValueAxis* self, intptr_t slot) {
    void (*slotFunc)(QLogValueAxis*, int) = reinterpret_cast<void (*)(QLogValueAxis*, int)>(slot);
    QLogValueAxis::connect(self, &QLogValueAxis::minorTickCountChanged, [self, slotFunc](int minorTickCount) {
	slotFunc(self, minorTickCount);
    });
}

libqt_string QLogValueAxis_Tr2(const char* s, const char* c) {
	QString _ret = QLogValueAxis::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QLogValueAxis_Tr3(const char* s, const char* c, int n) {
	QString _ret = QLogValueAxis::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
int QLogValueAxis_QBaseType(const QLogValueAxis* self) {
	auto* vqlogvalueaxis = dynamic_cast<const VirtualQLogValueAxis*>(self);
	if (vqlogvalueaxis && vqlogvalueaxis->isVirtualQLogValueAxis) {
vqlogvalueaxis->setQLogValueAxis_Type_IsBase(true);
	return vqlogvalueaxis->type();
}
}

// Auxiliary method to allow providing re-implementation
void QLogValueAxis_OnType(const QLogValueAxis* self, intptr_t slot) {
	auto* vqlogvalueaxis = dynamic_cast<const VirtualQLogValueAxis*>(self);
	if (vqlogvalueaxis && vqlogvalueaxis->isVirtualQLogValueAxis) {
vqlogvalueaxis->setQLogValueAxis_Type_Callback(reinterpret_cast<VirtualQLogValueAxis::QLogValueAxis_Type_Callback>(slot));
}
}

void QLogValueAxis_Delete(QLogValueAxis* self) {
    delete self;
}

File: binding/include/charts/libqlogvalueaxis.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qlogvalueaxis.html
pub const qlogvalueaxis = struct {

    /// New constructs a new QLogValueAxis object.
    pub fn New() QtC.QLogValueAxis {
        return qtc.QLogValueAxis_new();
    }


    /// New2 constructs a new QLogValueAxis object.
    pub fn New2(parent: ?*anyopaque) QtC.QLogValueAxis {
        return qtc.QLogValueAxis_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QLogValueAxis_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qlogvalueaxis.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Type(self: ?*anyopaque, ) i32 {
        return qtc.QLogValueAxis_Type(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnType(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i32) void {
        qtc.QLogValueAxis_OnType(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseType(self: ?*anyopaque, ) i32 {
        return qtc.QLogValueAxis_QBaseType(@ptrCast(self));
    }

    pub fn SetMin(self: ?*anyopaque, min: f64) void {
        qtc.QLogValueAxis_SetMin(@ptrCast(self), min);
    }

    pub fn Min(self: ?*anyopaque, ) f64 {
        return qtc.QLogValueAxis_Min(@ptrCast(self));
    }

    pub fn SetMax(self: ?*anyopaque, max: f64) void {
        qtc.QLogValueAxis_SetMax(@ptrCast(self), max);
    }

    pub fn Max(self: ?*anyopaque, ) f64 {
        return qtc.QLogValueAxis_Max(@ptrCast(self));
    }

    pub fn SetRange(self: ?*anyopaque, min: f64, max: f64) void {
        qtc.QLogValueAxis_SetRange(@ptrCast(self), min, max);
    }

    pub fn SetLabelFormat(self: ?*anyopaque, format: []const u8) void {
        const format_str = qtc.libqt_string{
    .len = format.len,
    .data = format.ptr,
};
qtc.QLogValueAxis_SetLabelFormat(@ptrCast(self), format_str);
    }

    pub fn LabelFormat(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QLogValueAxis_LabelFormat(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qlogvalueaxis.LabelFormat: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetBase(self: ?*anyopaque, base: f64) void {
        qtc.QLogValueAxis_SetBase(@ptrCast(self), base);
    }

    pub fn Base(self: ?*anyopaque, ) f64 {
        return qtc.QLogValueAxis_Base(@ptrCast(self));
    }

    pub fn TickCount(self: ?*anyopaque, ) i32 {
        return qtc.QLogValueAxis_TickCount(@ptrCast(self));
    }

    pub fn SetMinorTickCount(self: ?*anyopaque, minorTickCount: i32) void {
        qtc.QLogValueAxis_SetMinorTickCount(@ptrCast(self), minorTickCount);
    }

    pub fn MinorTickCount(self: ?*anyopaque, ) i32 {
        return qtc.QLogValueAxis_MinorTickCount(@ptrCast(self));
    }

    pub fn MinChanged(self: ?*anyopaque, min: f64) void {
        qtc.QLogValueAxis_MinChanged(@ptrCast(self), min);
    }

    pub fn OnMinChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, f64) callconv(.c) void) void {
        qtc.QLogValueAxis_Connect_MinChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn MaxChanged(self: ?*anyopaque, max: f64) void {
        qtc.QLogValueAxis_MaxChanged(@ptrCast(self), max);
    }

    pub fn OnMaxChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, f64) callconv(.c) void) void {
        qtc.QLogValueAxis_Connect_MaxChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn RangeChanged(self: ?*anyopaque, min: f64, max: f64) void {
        qtc.QLogValueAxis_RangeChanged(@ptrCast(self), min, max);
    }

    pub fn OnRangeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, f64, f64) callconv(.c) void) void {
        qtc.QLogValueAxis_Connect_RangeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn LabelFormatChanged(self: ?*anyopaque, format: []const u8) void {
        const format_str = qtc.libqt_string{
    .len = format.len,
    .data = format.ptr,
};
qtc.QLogValueAxis_LabelFormatChanged(@ptrCast(self), format_str);
    }

    pub fn OnLabelFormatChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8) callconv(.c) void) void {
        qtc.QLogValueAxis_Connect_LabelFormatChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn BaseChanged(self: ?*anyopaque, base: f64) void {
        qtc.QLogValueAxis_BaseChanged(@ptrCast(self), base);
    }

    pub fn OnBaseChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, f64) callconv(.c) void) void {
        qtc.QLogValueAxis_Connect_BaseChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn TickCountChanged(self: ?*anyopaque, tickCount: i32) void {
        qtc.QLogValueAxis_TickCountChanged(@ptrCast(self), tickCount);
    }

    pub fn OnTickCountChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QLogValueAxis_Connect_TickCountChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn MinorTickCountChanged(self: ?*anyopaque, minorTickCount: i32) void {
        qtc.QLogValueAxis_MinorTickCountChanged(@ptrCast(self), minorTickCount);
    }

    pub fn OnMinorTickCountChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QLogValueAxis_Connect_MinorTickCountChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QLogValueAxis_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qlogvalueaxis.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QLogValueAxis_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qlogvalueaxis.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QLogValueAxis_Delete(@ptrCast(self));
    }
};
